tutorial/tests/testsuite.py

```diff
@@ -0,0 +1,79 @@
+import io
+from contextlib import redirect_stdout
+from typing import Callable
+
+import pytest
+import ipynbname
+from IPython.core.magic import register_cell_magic
+from IPython.display import HTML, display
+
+class FunctionInjectionPlugin:
+    """A class to inject a function to test"""
+
+    def __init__(self, function_to_test: Callable) -> None:
+        self.function_to_test = function_to_test
+
+    def pytest_generate_tests(self, metafunc: pytest.Metafunc) -> None:
+        """Override the abstract `function_to_test` fixture function"""
+        if "function_to_test" in metafunc.fixturenames:
+            metafunc.parametrize("function_to_test", [self.function_to_test])
+
+@pytest.fixture
+def function_to_test():
+    """Function to test, overriden at runtime by the cell magic"""
+
+@register_cell_magic
+def test_functions_from_cell(line, cell):
+    # Execute the cell code in a new namespace
+
+    # Get the path to the current notebook file
+    nbname = ipynbname.name()
+
+    local_ns = {}
+    exec(cell, local_ns)
```

Comment:
This is the only change I would question: it suppresses any IPython output from the cell that's not a side-effect. So, either we force users to **only** define the functions and avoid any other kind of statement, or we must go back to `shell.run_cell(cell)` – but we need a class for that, not simply the `@register_cell_magic` decorator.

Reply:
I agree 👍